dev/NumberBox/NumberBoxParser.cpp

```diff
@@ -0,0 +1,295 @@
+{
+    this->type = t;
+    this->str = s;
+}
+
+// Pass directly
+MathToken::MathToken(TokenType t, std::wstring s)
+{
+    this->type = t;
+    this->str = s;
+}
+
+// Empty Token
+MathToken::MathToken()
+{
+    this->type = MathToken::TokenType::EOFToken;
+    this->str = L"";
+}
+
+MathTokenizer::MathTokenizer(std::wstring input)
+{
+    m_inputString = input;
+    m_inputLength = static_cast<int>(m_inputString.size());
+    m_lastToken = MathToken(MathToken::TokenType::EOFToken, L"");
+}
+
+MathToken MathTokenizer::GetToken()
+{
+    // Skip Whitespaces
+    SkipWhiteSpace();
+
+    if (m_index >= m_inputLength)
+    {
+        return MathToken(MathToken::TokenType::EOFToken, L"");
+    }
+ 
+    std::wstringstream ss;
+    ss << m_inputString[m_index];
+
+    // Check for a negative number. Next Token must be numeric and previous token must not be numeric.
+    if (m_inputString[m_index] == '-')
+    {
+        
+        if (PeekNextToken().type == MathToken::TokenType::Numeric && m_lastToken.type == MathToken::TokenType::Operator)
+        {
+            // Read next token so that the string is parsed as a numeric token.
+            m_index++;
+            SkipWhiteSpace();
+            ss << m_inputString[m_index];
+        }
+    }
+
+    // Return token that is +,-,*,^,/
+    if (IsOperator(ss.str()))
+    {
+        m_index++;
+        m_lastToken = MathToken(MathToken::TokenType::Operator, ss.str());
+    }
+    // Return parenth token
+    else if (m_inputString[m_index] == '(')
+    {
+        m_index++;
+        m_lastToken = MathToken(MathToken::TokenType::ParenLeft, ss.str());;
+    }
+    else if (m_inputString[m_index] == ')')
+    {
+        m_index++;
+        m_lastToken = MathToken(MathToken::TokenType::ParenRight, ss.str());
+    }
+    // Numeric Tokens
+    else if (IsNumeric(std::wstring_view(ss.str())))
+    {
+        // Keep Reading until all of number is parsed
+        while (IsNumeric(std::wstring_view(ss.str())) && m_index < m_inputLength)
+        {
+            ss << m_inputString[++m_index];
+        }
+        m_lastToken = MathToken(MathToken::TokenType::Numeric, ss.str().substr(0, ss.str().size() - 1));
+    }
+    else
+    {
+        m_index++;
+        m_lastToken = MathToken(MathToken::TokenType::Invalid, ss.str());
+    }
+    return m_lastToken;
+
+}
+
+// Peek at the next token that will be returned. Used for negative checks.
+MathToken MathTokenizer::PeekNextToken()
+{
+    // Save statics that need to be restored
+    MathToken oldLastToken = m_lastToken;
+    int oldIndex = m_index++;
+    MathToken nextToken = GetToken();
+    // Restore statics
+    m_index = oldIndex;
+    m_lastToken = oldLastToken;
+    return nextToken;
+}
+
+void MathTokenizer::SkipWhiteSpace()
+{
+    while (m_inputString[m_index] == ' ' && m_index <= m_inputLength)
+    {
+        m_index++;
+    }
+}
+
+bool MathTokenizer::IsNumeric(std::wstring_view in)
+{
+    std::wregex r(L"^-?(\\d?)+(\\.)?(\\d?)+?$");
```

Comment:
regex is an expensive way to check this -- also remember that numeric is locale-dependent so we should figure out how to use the number parser here. It might be that digits and period aren't numeric in the current locale.

Reply:
Ah you're right, I can probably just use a DecimalFormatter for this - it just has some less predictable behavior but if I can track down the indexing crashes it causes i'll swap it over to that.